File: vm/src/main/java/com/libreshockwave/vm/opcode/ExecutionContext.java
package com.libreshockwave.vm.opcode;

import com.libreshockwave.chunks.ScriptChunk;
import com.libreshockwave.vm.Datum;
import com.libreshockwave.vm.HandlerRef;
import com.libreshockwave.vm.LingoException;
import com.libreshockwave.vm.Scope;
import com.libreshockwave.vm.TraceListener;
import com.libreshockwave.vm.builtin.BuiltinRegistry;

import java.util.ArrayList;
import java.util.List;

/**
 * Execution context passed to opcode handlers.
 * Provides access to the current scope and VM utilities needed during instruction execution.
 */
public final class ExecutionContext {

    private final Scope scope;
    private final ScriptChunk.Handler.Instruction instruction;
    private final int argument;
    private final BuiltinRegistry builtins;
    private final TraceListener traceListener;

    // Callbacks to LingoVM
    private final HandlerExecutor handlerExecutor;
    private final HandlerFinder handlerFinder;
    private final GlobalAccessor globalAccessor;
    private final BuiltinInvoker builtinInvoker;

    public ExecutionContext(
            Scope scope,
            ScriptChunk.Handler.Instruction instruction,
            BuiltinRegistry builtins,
            TraceListener traceListener,
            HandlerExecutor handlerExecutor,
            HandlerFinder handlerFinder,
            GlobalAccessor globalAccessor,
            BuiltinInvoker builtinInvoker) {
        this.scope = scope;
        this.instruction = instruction;
        this.argument = instruction.argument();
        this.builtins = builtins;
        this.traceListener = traceListener;
        this.handlerExecutor = handlerExecutor;
        this.handlerFinder = handlerFinder;
        this.globalAccessor = globalAccessor;
        this.builtinInvoker = builtinInvoker;
    }

    // Scope access

    public Scope getScope() {
        return scope;
    }

    public int getArgument() {
        return argument;
    }

    public ScriptChunk.Handler.Instruction getInstruction() {
        return instruction;
    }

    // Stack operations (delegated to scope)

    public void push(Datum value) {
        scope.push(value);
    }

    public Datum pop() {
        return scope.pop();
    }

    public Datum peek() {
        return scope.peek();
    }

    public Datum peek(int depth) {
        return scope.peek(depth);
    }

    public void swap() {
        scope.swap();
    }

    // Local/param access

    public Datum getLocal(int index) {
        return scope.getLocal(index);
    }

    public void setLocal(int index, Datum value) {
        scope.setLocal(index, value);
        if (traceListener != null) {
            traceListener.onVariableSet("local", "local" + index, value);
        }
    }

    public Datum getParam(int index) {
        return scope.getParam(index);
    }

    // Global access

    public Datum getGlobal(String name) {
        return globalAccessor.getGlobal(name);
    }

    public void setGlobal(String name, Datum value) {
        globalAccessor.setGlobal(name, value);
        if (traceListener != null) {
            traceListener.onVariableSet("global", name, value);
        }
    }

    // Return handling

    public void setReturnValue(Datum value) {
        scope.setReturnValue(value);
    }

    public void setReturned(boolean returned) {
        scope.setReturned(returned);
    }

    // Name resolution - delegates to ScriptChunk which uses the correct cast lib's names

    public String resolveName(int nameId) {
        return getScript().resolveName(nameId);
    }

    // Script/handler access

    public ScriptChunk getScript() {
        return scope.getScript();
    }

    public ScriptChunk.Handler getHandler() {
        return scope.getHandler();
    }

    public Datum getReceiver() {
        return scope.getReceiver();
    }

    public List<ScriptChunk.LiteralEntry> getLiterals() {
        return scope.getScript().literals();
    }

    // Jump handling

    public void jumpTo(int targetOffset) {
        int targetIndex = scope.getHandler().getInstructionIndex(targetOffset);
        if (targetIndex >= 0) {
            scope.setBytecodeIndex(targetIndex);
        }
    }

    public int getInstructionOffset() {
        return instruction.offset();
    }

    // Function calls

    public ScriptChunk.Handler findLocalHandler(int vectorPos) {
        for (ScriptChunk.Handler handler : getScript().handlers()) {
            if (handler.handlerVectorPos() == vectorPos) {
                return handler;
            }
        }
        return null;
    }

    public HandlerRef findHandler(String handlerName) {
        return handlerFinder.findHandler(handlerName);
    }

    public Datum executeHandler(ScriptChunk script, ScriptChunk.Handler handler,
                                List<Datum> args, Datum receiver) {
        return handlerExecutor.executeHandler(script, handler, args, receiver);
    }

    public boolean isBuiltin(String name) {
        return builtins.contains(name);
    }

    public Datum invokeBuiltin(String name, List<Datum> args) {
        return builtinInvoker.invoke(name, args);
    }

    public BuiltinRegistry getBuiltins() {
        return builtins;
    }

    // Helper to pop multiple args in reverse order

    public List<Datum> popArgs(int count) {
        List<Datum> args = new ArrayList<>();
        for (int i = 0; i < count; i++) {
            args.add(0, pop());
        }
        return args;
    }

    // Property tracing

    public void tracePropertySet(String propName, Datum value) {
        if (traceListener != null) {
            traceListener.onVariableSet("property", "me." + propName, value);
        }
    }

    // Exception helper

    public LingoException error(String message) {
        return new LingoException(message);
    }

    // Functional interfaces for callbacks

    @FunctionalInterface
    public interface HandlerExecutor {
        Datum executeHandler(ScriptChunk script, ScriptChunk.Handler handler,
                             List<Datum> args, Datum receiver);
    }

    @FunctionalInterface
    public interface HandlerFinder {
        HandlerRef findHandler(String name);
    }

    public interface GlobalAccessor {
        Datum getGlobal(String name);
        void setGlobal(String name, Datum value);
    }

    @FunctionalInterface
    public interface BuiltinInvoker {
        Datum invoke(String name, List<Datum> args);
    }
}

File: cpp/src/lingo/vm/ExecutionContext.cpp
#include "libreshockwave/lingo/vm/ExecutionContext.hpp"

#include <limits>
#include <string>
#include <utility>

namespace libreshockwave::lingo::vm {

namespace {

void resetBuiltinControlFlow(builtin::BuiltinContext& context) {
    context.returned = false;
    context.aborted = false;
    context.returnValue = Datum::voidValue();
}

std::optional<Datum> consumeBuiltinReturnValue(builtin::BuiltinContext& context) {
    const bool returned = context.returned || context.aborted;
    Datum returnValue = returned ? context.returnValue : Datum::voidValue();
    resetBuiltinControlFlow(context);
    if (!returned) {
        return std::nullopt;
    }
    return std::move(returnValue);
}

} // namespace

ExecutionContext::ExecutionContext(Scope& scope,
                                   chunks::ScriptChunk::Instruction instruction,
                                   builtin::BuiltinRegistry* builtins,
                                   builtin::BuiltinContext* builtinContext,
                                   Callbacks callbacks,
                                   int variableMultiplier,
                                   bool instructionTraceEnabled)
    : scope_(&scope),
      instruction_(instruction),
      argument_(instruction.argument),
      variableMultiplier_(variableMultiplier > 0 ? variableMultiplier : 1),
      instructionTraceEnabled_(instructionTraceEnabled),
      builtins_(builtins),
      builtinContext_(builtinContext),
      callbacks_(std::move(callbacks)) {}

std::vector<Datum> ExecutionContext::popArgs(int count) {
    if (count <= 0) {
        return {};
    }
    std::vector<Datum> args(static_cast<std::size_t>(count));
    for (int index = count - 1; index >= 0; --index) {
        args[static_cast<std::size_t>(index)] = pop();
    }
    return args;
}

Datum ExecutionContext::getLocal(int index) const {
    return scope_->getLocal(index);
}

void ExecutionContext::setLocal(int index, Datum value) {
    if (!callbacks_.variableSetListener) {
        scope_->setLocal(index, std::move(value));
        return;
    }
    const Datum tracedValue = value;
    scope_->setLocal(index, std::move(value));
    callbacks_.variableSetListener("local", "local" + std::to_string(index), tracedValue);
}

Datum ExecutionContext::getParam(int index) const {
    return scope_->getParam(index);
}

void ExecutionContext::setParam(int index, Datum value) {
    if (!callbacks_.variableSetListener) {
        scope_->setParam(index, std::move(value));
        return;
    }
    const Datum tracedValue = value;
    scope_->setParam(index, std::move(value));
    callbacks_.variableSetListener("param", "param" + std::to_string(index), tracedValue);
}

Datum ExecutionContext::getGlobal(std::string_view name) const {
    if (callbacks_.globalGetter) {
        return callbacks_.globalGetter(name);
    }
    return Datum::voidValue();
}

void ExecutionContext::setGlobal(std::string_view name, Datum value) {
    if (!callbacks_.variableSetListener) {
        if (callbacks_.globalSetter) {
            callbacks_.globalSetter(name, std::move(value));
        }
        return;
    }
    const Datum tracedValue = value;
    if (callbacks_.globalSetter) {
        callbacks_.globalSetter(name, std::move(value));
    }
    callbacks_.variableSetListener("global", name, tracedValue);
}

void ExecutionContext::setErrorState(bool errorState) {
    if (callbacks_.errorStateSetter) {
        callbacks_.errorStateSetter(errorState);
    }
}

void ExecutionContext::tracePropertySet(std::string_view propName, const Datum& value) const {
    if (callbacks_.variableSetListener) {
        callbacks_.variableSetListener("property", "me." + std::string(propName), value);
    }
}

void ExecutionContext::jumpTo(int targetOffset) {
    int targetIndex = -1;
    if (targetOffset == cachedJumpOffset_) {
        targetIndex = cachedJumpIndex_;
    } else {
        targetIndex = scope_->handler().getInstructionIndex(targetOffset);
        cachedJumpOffset_ = targetOffset;
        cachedJumpIndex_ = targetIndex;
    }
    if (targetIndex >= 0) {
        scope_->setBytecodeIndex(targetIndex);
    }
}

const chunks::ScriptChunk::Handler* ExecutionContext::findLocalHandler(int index) const {
    const auto* script = scope_->script();
    if (script == nullptr || index < 0 || index >= static_cast<int>(script->handlers().size())) {
        return nullptr;
    }
    return &script->handlers()[static_cast<std::size_t>(index)];
}

const std::vector<chunks::ScriptChunk::LiteralEntry>& ExecutionContext::literals() const {
    static const std::vector<chunks::ScriptChunk::LiteralEntry> empty;
    const auto* script = scope_->script();
    return script != nullptr ? script->literals() : empty;
}

const std::string& ExecutionContext::resolveNameRef(int nameId) const {
    if (nameId == cachedResolvedNameId_ && cachedResolvedName_ != nullptr) {
        return *cachedResolvedName_;
    }
    if (const auto found = resolvedNames_.find(nameId); found != resolvedNames_.end()) {
        cachedResolvedNameId_ = nameId;
        cachedResolvedName_ = &found->second;
        return found->second;
    }
    std::string resolved;
    if (callbacks_.nameResolver) {
        resolved = callbacks_.nameResolver(nameId);
    } else {
        const auto* script = scope_->script();
        resolved = script != nullptr ? script->resolveName(nameId, nullptr) : "#" + std::to_string(nameId);
    }
    auto [inserted, _] = resolvedNames_.emplace(nameId, std::move(resolved));
    cachedResolvedNameId_ = nameId;
    cachedResolvedName_ = &inserted->second;
    return inserted->second;
}

std::string ExecutionContext::resolveName(int nameId) const {
    return resolveNameRef(nameId);
}

std::optional<HandlerRef> ExecutionContext::findHandler(std::string_view name) const {
    if (callbacks_.handlerFinder) {
        return callbacks_.handlerFinder(name);
    }
    return std::nullopt;
}

Datum ExecutionContext::executeHandler(const chunks::ScriptChunk& script,
                                       const chunks::ScriptChunk::Handler& handler,
                                       const std::vector<Datum>& args,
                                       const Datum& receiver) const {
    return executeHandler(script, handler, std::span<const Datum>(args), receiver);
}

Datum ExecutionContext::executeHandler(const chunks::ScriptChunk& script,
                                       const chunks::ScriptChunk::Handler& handler,
                                       std::span<const Datum> args,
                                       const Datum& receiver) const {
    return executeHandler(HandlerRef{&script, &handler}, args, receiver);
}

Datum ExecutionContext::executeHandler(const HandlerRef& handler,
                                       const std::vector<Datum>& args,
                                       const Datum& receiver) const {
    return executeHandler(handler, std::span<const Datum>(args), receiver);
}

Datum ExecutionContext::executeHandler(const HandlerRef& handler,
                                       std::span<const Datum> args,
                                       const Datum& receiver) const {
    if (callbacks_.handlerRefSpanExecutor) {
        return callbacks_.handlerRefSpanExecutor(handler, args, receiver);
    }
    const std::vector<Datum> argVector(args.begin(), args.end());
    if (callbacks_.handlerRefExecutor) {
        return callbacks_.handlerRefExecutor(handler, argVector, receiver);
    }
    if (callbacks_.handlerExecutor) {
        if (handler.script == nullptr || handler.handler == nullptr) {
            return Datum::voidValue();
        }
        return callbacks_.handlerExecutor(*handler.script, *handler.handler, argVector, receiver);
    }
    return Datum::voidValue();
}

bool ExecutionContext::isBuiltin(std::string_view name) const {
    return builtins_ != nullptr && builtins_->contains(name);
}

Datum ExecutionContext::invokeBuiltin(std::string_view name, const std::vector<Datum>& args) {
    if (callbacks_.builtinInvoker) {
        return callbacks_.builtinInvoker(name, args);
    }
    if (builtins_ == nullptr) {
        return Datum::voidValue();
    }
    builtin::BuiltinContext fallback;
    auto& context = builtinContext_ != nullptr ? *builtinContext_ : fallback;
    resetBuiltinControlFlow(context);
    Datum result = builtins_->invoke(name, context, args);
    if (auto returnValue = consumeBuiltinReturnValue(context)) {
        scope_->setReturnValue(std::move(*returnValue));
    }
    return result;
}

std::optional<Datum> ExecutionContext::invokeBuiltinIfPresent(std::string_view name, const std::vector<Datum>& args) {
    if (builtins_ == nullptr) {
        return std::nullopt;
    }
    builtin::BuiltinContext fallback;
    auto& context = builtinContext_ != nullptr ? *builtinContext_ : fallback;
    resetBuiltinControlFlow(context);
    auto result = builtins_->invokeIfPresent(name, context, args);
    if (auto returnValue = consumeBuiltinReturnValue(context)) {
        scope_->setReturnValue(std::move(*returnValue));
    }
    return result;
}

std::string ExecutionContext::formatCallStack() const {
    if (callbacks_.callStackFormatter) {
        return callbacks_.callStackFormatter();
    }
    return "(no call stack available)";
}

LingoException ExecutionContext::error(const std::string& message) const {
    return LingoException(message);
}

} // namespace libreshockwave::lingo::vm
